### ISSUE TITLE
`Object.as_json` is too optimistic and causes errors in 3rd party code.

### ISSUE THREAD

**Issue by ioquatix** [rails/rails#51626]
This monkey patch is dangerous:

https://github.com/rails/rails/blob/82054e8eb76a72ded0abee83019e60c19f552a8b/activesupport/lib/active_support/core_ext/object/json.rb#L58-L66

Specifically, I've seen this issue crop up a number of times. A casual search yields issues that appear to be related:

- https://github.com/socketry/async-http/issues/157
- https://github.com/socketry/async-http/issues/156
- https://github.com/rails/rails/issues/40781
- https://github.com/rails/rails/issues/16112
- https://github.com/rails/rails/issues/51202
- https://github.com/rails/rails/issues/43040
- https://github.com/rails/rails/issues/28996
- https://github.com/rails/rails/issues/31313
- https://github.com/rails/rails/issues/11631

I don't think `as_json` should be monkey patched into `Object` like this. The default behaviour of using `to_s` is safer. Alternatively, objects that refer to themselves (circular references) should be handled correctly.

**Comment by ioquatix** [comment]
I guess it was introduced here: https://github.com/rails/rails/commit/e567a5eb1afe1ac38f1da37f1c1e3922bbf79d2a

cc @jeremy you have worked on that code quite a bit, do you have any opinion?

**Comment by ioquatix** [comment]
A safer default might be this:

```ruby
 class Object 
   def as_json(options = nil) # :nodoc: 
     if respond_to?(:to_hash) 
       to_hash.as_json(options) 
     else 
       self.to_s
     end 
   end 
 end 
```

**Comment by Earlopain** [comment]
I've also seen this reported to sidekiq a few times:
* https://github.com/sidekiq/sidekiq/issues/6000
* https://github.com/sidekiq/sidekiq/issues/5920
* https://github.com/sidekiq/sidekiq/issues/5858 (maybe?)

This was eventually fixed by implementing a custom `to_hash` in https://github.com/sidekiq/sidekiq/pull/6198

**Comment by ioquatix** [comment]
The problem with `to_hash` is it makes an object implicitly convertible (equivalent to) a Hash. It would have made more sense to use `to_h` (explicitly convertible).

See <https://zverok.space/blog/2016-01-18-implicit-vs-expicit.html> for a good summary of the differences.

**Comment by ioquatix** [comment]
This is still an issue.

**Comment by ioquatix** [comment]
This is still an issue.

**Comment by jrmhaig** [comment]
If it helps I have a minimal example of the issue we are seeing: https://github.com/jrmhaig/rails_issue_51626_example

In summary;
* We are passing the form builder instance to a partial
* We use the [`logstasher`](https://rubygems.org/gems/logstasher) gem with the debug log level in development
* This works in Rails 7.0 and results in a 'stack level too deep' error in Rails 7.1 when `logstasher` attempts to log details of the partial

**Comment by benoittgt** [comment]
> I hate that I need to monkey-patch Object (maybe I can find the offending class that is lacking a to_hash method instead)

This what I did. I monitored classes that were not having `to_hash` in ActiveSupport `Object#as_json`, then implement a fake `to_hash` to see the impact, and it pass. To try add into [simple form source code](https://github.com/heartcombo/simple_form/blob/c2c7faf3e532072e1ef511a97d677ae0e7c38957/lib/simple_form/form_builder.rb#L47) something like `def to_hash; {a: 1}; end`. Then check.

I my case, the "to_json" is requested by logstasher. I think logstasher could be smarter.

**Comment by byroot** [comment]
For better or for worse, that's how the underlying JSON implementation (and all JSON implementations I know about) deal with recursion. `JSON.generate` will raise on recursive data structures, so will `OJ.dump`, `YAJL.dump`, etc.

Detecting recursion means keeping a set of all the traversed references, which is prohibitively costly.

I don't see any way to action this issue, and no, calling `to_s` isn't better, it's way worse because while it won't raise, it will accept anything and will never be what you want.

That being said, the fallback on `instance_values.as_json(options)` definitely isn't great though, this sort of automatic conversion to JSON is a recipe for disaster, and it would be good to deprecate it in favor of a  simple `NoMethodError`.

As for `to_hash` vs `to_h`, `to_hash` is correct here IMO.

**Comment by ioquatix** [comment]
I agree `instance_values.as_json(options)` is terrible as it spills the guts of the object in a way that was probably never intended, and I believe this probably accounts for the majority of issues - at least IIRC the ones I experienced.
